File: backend/logic/preset/preset.go
package preset

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"

	"connectrpc.com/connect"
	"github.com/ling/muse/common/constant"
	"github.com/ling/muse/common/convert"
	"github.com/ling/muse/common/errs"
	"github.com/ling/muse/common/jwt"
	"github.com/ling/muse/entity"
	"github.com/ling/muse/entity/sillytavern"
	pb "github.com/ling/muse/gen/muse"
	"github.com/ling/muse/repo/cache"
	"github.com/ling/muse/repo/database"
	log "github.com/sirupsen/logrus"
)

type presetImpl struct {
	presetRepo    database.PresetRepository
	regexRuleRepo database.RegexRuleRepository
}

func newPreset() *presetImpl {
	return &presetImpl{
		presetRepo:    database.NewPresetRepo(),
		regexRuleRepo: database.NewRegexRuleRepo(),
	}
}

func (p *presetImpl) ListPresets(ctx context.Context, req *pb.ListPresetsRequest) (*pb.ListPresetsResponse, error) {
	// 获取并规范化分页参数
	page, pageSize := constant.NormalizePagination(int(req.GetPage()), int(req.GetPageSize()))

	// 从数据库获取预设列表（不加载关联的PromptItems和RegexRules）
	userId := jwt.GetUserId(ctx)
	presets, presetProLen, total, err := p.presetRepo.List(ctx, userId, page, pageSize)
	if err != nil {
		return nil, err
	}

	// 转换为pb格式
	pbPresets := make([]*pb.PresetWithPromptLen, 0, len(presets))
	for i := range presets {
		pbPreset, _, _ := convert.PresetEntityToPb(presets[i])
		pbPresets = append(pbPresets, &pb.PresetWithPromptLen{
			Preset:    pbPreset,
			PromptLen: int32(presetProLen[i]),
		})
	}

	return &pb.ListPresetsResponse{
		Presets:  pbPresets,
		Total:    total,
		Page:     int32(page),
		PageSize: int32(pageSize),
	}, nil
}

func (p *presetImpl) GetPreset(ctx context.Context, req *pb.GetPresetRequest) (*pb.GetPresetResponse, error) {
	id := int(req.GetId())
	if id <= 0 {
		return nil, errs.NewStandard(connect.CodeInvalidArgument, errs.InvalidPresetID)
	}

	// 从数据库获取预设
	userId := jwt.GetUserId(ctx)
	preset, err := p.presetRepo.GetByID(ctx, id, userId)
	if err != nil {
		return nil, err
	}
	if preset == nil {
		return nil, errs.NewStandard(connect.CodeNotFound, errs.PresetNotFound)
	}

	pbPreset, pbPromptItem, pbRegex := convert.PresetEntityToPb(preset)
	return &pb.GetPresetResponse{
		Preset: &pb.PresetWithAll{
			Preset:      pbPreset,
			PromptItems: pbPromptItem,
			RegexRules:  pbRegex,
		},
	}, nil
}

func (p *presetImpl) CreatePreset(ctx context.Context, req *pb.CreatePresetRequest) (*pb.CreatePresetResponse, error) {
	// 参数校验
	name := strings.TrimSpace(req.GetName())
	if name == "" {
		return nil, errs.NewStandard(connect.CodeInvalidArgument, errs.EmptyPresetName)
	}

	// 构建预设实体
	userId := jwt.GetUserId(ctx)
	preset := &entity.Preset{
		UserID:           userId,
		Name:             name,
		Temperature:      req.GetTemperature(),
		TopP:             req.GetTopP(),
		TopK:             int(req.GetTopK()),
		MaxTokens:        int(req.GetMaxTokens()),
		FrequencyPenalty: req.GetFrequencyPenalty(),
		PresencePenalty:  req.GetPresencePenalty(),
		Version:          1,
	}

	// 保存到数据库
	if err := p.presetRepo.Create(ctx, preset); err != nil {
		return nil, err
	}

	// 创建关联的提示项
	for _, itemReq := range req.GetPromptItems() {
		item := &entity.PromptItem{
			PresetID:          preset.ID,
			Identifier:        itemReq.GetIdentifier(),
			Name:              strings.TrimSpace(itemReq.GetName()),
			Content:           itemReq.GetContent(),
			Role:              itemReq.GetRole(),
			IsEnabled:         itemReq.GetIsEnabled(),
			InjectionPosition: itemReq.GetInjectionPosition(),
			InjectionDepth:    int(itemReq.GetInjectionDepth()),
			ForbidOverrides:   itemReq.GetForbidOverrides(),
		}
		// 创建提示项，忽略单个失败继续创建其他项
		_ = p.presetRepo.CreatePromptItem(ctx, item)
	}

	return &pb.CreatePresetResponse{}, nil
}

func (p *presetImpl) UpdatePreset(ctx context.Context, req *pb.UpdatePresetRequest) (*pb.UpdatePresetResponse, error) {
	id := int(req.GetId())
	if id <= 0 {
		return nil, errs.NewStandard(connect.CodeInvalidArgument, errs.InvalidPresetID)
	}

	// 参数校验
	name := strings.TrimSpace(req.GetName())
	if name == "" {
		return nil, errs.NewStandard(connect.CodeInvalidArgument, errs.EmptyPresetName)
	}

	// 获取当前预设
	userId := jwt.GetUserId(ctx)
	preset, err := p.presetRepo.GetByID(ctx, id, userId)
	if err != nil {
		return nil, err
	}
	if preset == nil {
		return nil, errs.NewStandard(connect.CodeNotFound, errs.PresetNotFound)
	}

	// 检查版本号
	if int64(preset.Version) != req.GetVersion() {
		return nil, errs.NewStandard(connect.CodeAborted, errs.DataConflict)
	}

	// 更新预设字段
	preset.Name = name
	preset.Temperature = req.Temperature
	preset.TopP = req.TopP
	preset.TopK = int(req.TopK)
	preset.MaxTokens = int(req.MaxTokens)
	preset.FrequencyPenalty = req.FrequencyPenalty
	preset.PresencePenalty = req.PresencePenalty

	// 更新数据库
	if err = p.presetRepo.Update(ctx, preset); err != nil {
		return nil, err
	}

	// 使相关缓存失效
	cache.InvalidateCacheByPreset(int64(id))

	return &pb.UpdatePresetResponse{}, nil
}

func (p *presetImpl) DeletePreset(ctx context.Context, req *pb.DeletePresetRequest) (*pb.DeletePresetResponse, error) {
	id := int(req.GetId())
	if id <= 0 {
		return nil, errs.NewStandard(connect.CodeInvalidArgument, errs.InvalidPresetID)
	}

	// 使相关缓存失效（在删除之前）
	cache.InvalidateCacheByPreset(int64(id))

	// 删除预设（会级联删除关联的提示项）
	userId := jwt.GetUserId(ctx)
	if err := p.presetRepo.Delete(ctx, id, userId); err != nil {
		return nil, err
	}

	return &pb.DeletePresetResponse{}, nil
}

func (p *presetImpl) ListPromptItems(ctx context.Context, req *pb.ListPromptItemsRequest) (*pb.ListPromptItemsResponse, error) {
	presetID := int(req.GetPresetId())
	if presetID <= 0 {
		return nil, errs.NewStandard(connect.CodeInvalidArgument, errs.InvalidPresetID)
	}

	// 从数据库获取提示项列表
	items, err := p.presetRepo.ListPromptItems(ctx, presetID)
	if err != nil {
		return nil, err
	}

	// 转换为pb格式
	pbItems := make([]*pb.PromptItem, 0, len(items))
	for _, item := range items {
		pbItems = append(pbItems, convert.PromptItemEntityToPb(item))
	}

	return &pb.ListPromptItemsResponse{
		Items: pbItems,
	}, nil
}

func (p *presetImpl) AddPromptItem(ctx context.Context, req *pb.AddPromptItemRequest) (*pb.AddPromptItemResponse, error) {
	// 参数校验
	presetID := int(req.GetPresetId())
	if presetID <= 0 {
		return nil, errs.NewStandard(connect.CodeInvalidArgument, errs.InvalidPresetID)
	}

	name := strings.TrimSpace(req.GetName())
	if name == "" {
		return nil, errs.NewStandard(connect.CodeInvalidArgument, errs.EmptyPromptItemName)
	}

	// 构建提示项实体
	item := &entity.PromptItem{
		PresetID:          presetID,
		Identifier:        req.GetIdentifier(),
		Name:              name,
		Content:           req.GetContent(),
		Role:              req.GetRole(),
		IsEnabled:         req.GetIsEnabled(),
		InjectionPosition: req.GetInjectionPosition(),
		InjectionDepth:    int(req.GetInjectionDepth()),
		ForbidOverrides:   req.GetForbidOverrides(),
	}

	// 保存到数据库
	if err := p.presetRepo.CreatePromptItem(ctx, item); err != nil {
		return nil, err
	}

	// 重新获取完整数据
	fullItem, err := p.presetRepo.GetPromptItemByID(ctx, item.ID)
	if err != nil {
		return nil, err
	}

	return &pb.AddPromptItemResponse{
		Item: convert.PromptItemEntityToPb(fullItem),
	}, nil
}

func (p *presetImpl) UpdatePromptItem(ctx context.Context, req *pb.UpdatePromptItemRequest) (*pb.UpdatePromptItemResponse, error) {
	id := int(req.GetId())
	if id <= 0 {
		return nil, errs.NewStandard(connect.CodeInvalidArgument, errs.InvalidPromptItemID)
	}

	// 参数校验
	name := strings.TrimSpace(req.Name)
	if name == "" {
		return nil, errs.NewStandard(connect.CodeInvalidArgument, errs.EmptyPromptItemName)
	}

	// 获取当前提示项
	item, err := p.presetRepo.GetPromptItemByID(ctx, id)
	if err != nil {
		return nil, err
	}
	if item == nil {
		return nil, errs.NewStandard(connect.CodeNotFound, errs.PromptItemNotFound)
	}

	// 更新提示项字段
	item.Identifier = req.GetIdentifier()
	item.Name = name
	if req.Content != nil {
		item.Content = *req.Content
	}
	item.Role = req.Role
	item.IsEnabled = req.IsEnabled
	item.InjectionPosition = req.InjectionPosition
	item.InjectionDepth = int(req.InjectionDepth)
	item.ForbidOverrides = req.ForbidOverrides

	// 更新数据库
	if err := p.presetRepo.UpdatePromptItem(ctx, item); err != nil {
		return nil, err
	}

	// 使相关缓存失效（提示项修改会影响使用该预设的缓存）
	cache.InvalidateCacheByPreset(int64(item.PresetID))

	// 重新获取更新后的提示项
	updatedItem, err := p.presetRepo.GetPromptItemByID(ctx, id)
	if err != nil {
		return nil, err
	}

	return &pb.UpdatePromptItemResponse{
		Item: convert.PromptItemEntityToPb(updatedItem),
	}, nil
}

func (p *presetImpl) DeletePromptItem(ctx context.Context, req *pb.DeletePromptItemRequest) (*pb.DeletePromptItemResponse, error) {
	id := int(req.GetId())
	if id <= 0 {
		return nil, errs.NewStandard(connect.CodeInvalidArgument, errs.InvalidPromptItemID)
	}

	// 删除提示项
	if err := p.presetRepo.DeletePromptItem(ctx, id); err != nil {
		return nil, err
	}

	return &pb.DeletePromptItemResponse{}, nil
}

func (p *presetImpl) UpdatePromptItemsOrder(ctx context.Context, req *pb.UpdatePromptItemsOrderRequest) (*pb.UpdatePromptItemsOrderResponse, error) {
	presetID := int(req.GetPresetId())
	if presetID <= 0 {
		return nil, errs.NewStandard(connect.CodeInvalidArgument, errs.InvalidPresetID)
	}

	if req.SourceId <= 0 || req.DesId <= 0 || req.SortOperation == pb.SortOperation_OrderOperationUnspecified {
		return nil, errs.NewStandard(connect.CodeInvalidArgument, errs.EmptySortData)
	}

	userID := jwt.GetUserId(ctx)
	// 更新排序
	if err := p.presetRepo.UpdatePromptItemsOrder(ctx, presetID, userID, int(req.SourceId), int(req.DesId), req.SortOperation); err != nil {
		return nil, err
	}

	return &pb.UpdatePromptItemsOrderResponse{}, nil
}

func (p *presetImpl) SetActivePreset(ctx context.Context, req *pb.SetActivePresetRequest) (*pb.SetActivePresetResponse, error) {
	presetID := int(req.GetPresetId())
	if presetID <= 0 {
		return nil, errs.NewStandard(connect.CodeInvalidArgument, errs.InvalidPresetID)
	}
	if err := p.presetRepo.SetActivePreset(ctx, jwt.GetUserId(ctx), presetID); err != nil {
		return nil, err
	}
	return &pb.SetActivePresetResponse{}, nil
}

func (p *presetImpl) ImportPreset(ctx context.Context, req *pb.ImportPresetRequest) (*pb.ImportPresetResponse, error) {
	fileContent := req.GetFileContent()
	fileName := strings.TrimSpace(req.GetFileName())

	// 移除 UTF-8 BOM（如果存在）
	if len(fileContent) >= 3 && fileContent[0] == 0xEF && fileContent[1] == 0xBB && fileContent[2] == 0xBF {
		fileContent = fileContent[3:]
	}

	// 解析 SillyTavern 预设文件
	stPreset := &sillytavern.OpenAIPreset{}
	if err := json.Unmarshal(fileContent, stPreset); err != nil {
		return nil, errs.NewStandardf(connect.CodeInvalidArgument, "无效的预设文件: %v", err.Error())
	}

	// 生成预设名称：优先使用文件名，否则使用默认名称
	presetName := "导入的预设"
	if fileName != "" {
		// 移除文件扩展名
		presetName = strings.TrimSuffix(fileName, ".json")
	}

	// 使用 convert 包转换为 Muse 预设实体
	userId := jwt.GetUserId(ctx)
	preset := convert.STPresetToEntity(stPreset, userId, presetName)

	// 保存预设到数据库
	if err := p.presetRepo.Create(ctx, preset); err != nil {
		return nil, err
	}

	// 转换并保存提示项
	if len(stPreset.Prompts) > 0 {
		promptItems := orderPrompt(stPreset.Prompts, stPreset.PromptOrder)
		items := make([]*entity.PromptItem, 0, len(promptItems))
		for _, stPrompt := range promptItems {
			// 使用 convert 包转换提示项
			item := convert.STPromptToEntity(preset.ID, &stPrompt)
			items = append(items, item)
		}
		if err := p.presetRepo.BatchCreatePromptItem(ctx, items); err != nil {
			log.Errorf("failed to batch create prompt items: %v", err)
			return nil, err
		}
		log.Infof("batch created %d prompt items", len(items))
		// 排序提示项
		var pre *int = nil
		for i := 0; i < len(items); i++ {
			items[i].Pre = pre
			pre = &items[i].ID
			if i < len(items)-1 {
				items[i].Next = &items[i+1].ID
			}
			preLog := "nil"
			if items[i].Pre != nil {
				preLog = fmt.Sprintf("%d", *items[i].Pre)
			}
			nextLog := "nil"
			if items[i].Next != nil {
				nextLog = fmt.Sprintf("%d", *items[i].Next)
			}
			log.Debugf("当前提示项 ID: %d, Pre: %s, Next: %s", items[i].ID, preLog, nextLog)
		}
		if err := p.presetRepo.BatchUpdatePromptItemOrder(ctx, items); err != nil {
			log.Errorf("failed to batch update prompt item order: %v", err)
			return nil, err
		}
	}

	// TODO 导入预设内嵌的正则脚本
	if len(stPreset.Extensions.RegexScripts) > 0 {
		regexRules := make([]*entity.RegexRule, 0, len(stPreset.Extensions.RegexScripts))
		for _, stScript := range stPreset.Extensions.RegexScripts {
			// 使用 convert 包转换正则脚本，关联到新创建的预设（characterID=0 表示非角色范围正则）
			rule := convert.STRegexToEntity(&stScript, userId)
			regexRules = append(regexRules, rule)
		}

		// 批量创建正则规则
		if err := p.regexRuleRepo.BatchCreate(ctx, regexRules); err != nil {
			log.Errorf("failed to batch create regex rules: %v", err)
			return nil, errs.NewStandardf(errs.Code(err), "导入预设时批量创建正则规则失败: %v", err.Error())
		}
	}

	// 重新获取完整预设数据（包含关联的提示项）
	fullPreset, err := p.presetRepo.GetByID(ctx, preset.ID, userId)
	if err != nil {
		return nil, err
	}

	pbPreset, pbPromptItem, pbRegex := convert.PresetEntityToPb(fullPreset)
	return &pb.ImportPresetResponse{
		Preset: &pb.PresetWithAll{
			Preset:      pbPreset,
			PromptItems: pbPromptItem,
			RegexRules:  pbRegex,
		},
	}, nil
}

func (p *presetImpl) ExportPreset(ctx context.Context, req *pb.ExportPresetRequest) (*pb.ExportPresetResponse, error) {
	//TODO implement me
	panic("implement me")
}

// orderPrompt 构建提示词顺序映射
func orderPrompt(prompt []sillytavern.PresetPromptItem,
	promptOrder []sillytavern.PromptOrderItem) []sillytavern.PresetPromptItem {

	promptMap := make(map[string]*sillytavern.PresetPromptItem, len(prompt))
	for _, item := range prompt {
		promptMap[item.Identifier] = &item
	}
	orderPrompts := make([]sillytavern.PresetPromptItem, 0, len(prompt))

	const defaultCharacterID = 100001

	for _, order := range promptOrder {
		if order.CharacterID != defaultCharacterID {
			continue
		}
		for _, item := range order.Order {
			if promptItem, exist := promptMap[item.Identifier]; exist {
				promptItem.Enabled = item.Enabled // 注意这里特别坑, Enable放在了顺序这里, 所以得手动赋值到外层结构体
				orderPrompts = append(orderPrompts, *promptItem)
			}
		}
		break
	}
	// 如果没有找到默认角色，使用第一个配置
	if len(orderPrompts) == 0 && len(promptOrder) > 0 {
		log.Info("没有找到默认角色，使用第一个配置")
		for _, item := range promptOrder[0].Order {
			if promptItem, exist := promptMap[item.Identifier]; exist {
				promptItem.Enabled = item.Enabled // 注意这里特别坑, Enable放在了顺序这里, 所以得手动赋值到外层结构体
				orderPrompts = append(orderPrompts, *promptItem)
			}
		}
	}
	return orderPrompts
}

File: frontend/src/stores/preset.ts
import { defineStore } from 'pinia';
import { ref, computed } from 'vue';
import { presetClient } from '@/api/client';
import type { PromptItem, PresetWithAll, PresetWithPromptLen } from '@/gen/muse/preset_pb';
import type { Role, InjectionPosition, PromptItemIdentifier } from '@/gen/muse/common_pb';
import { DEFAULT_PAGE_NUM, DEFAULT_PAGE_SIZE, FETCH_ALL_PAGE_SIZE } from '@/utils/constants';

export const usePresetStore = defineStore('preset', () => {
  // =====================
  // 状态
  // =====================

  // 预设列表
  const presets = ref<PresetWithPromptLen[]>([]);
  // 当前选中的预设
  const selectedPreset = ref<PresetWithAll | null>(null);
  // 当前预设的提示项列表
  const promptItems = ref<PromptItem[]>([]);
  // 加载状态
  const loading = ref(false);
  const loadingMore = ref(false);
  // 搜索关键词
  const searchQuery = ref('');
  // 分页状态
  const pagination = ref({
    page: DEFAULT_PAGE_NUM,
    pageSize: DEFAULT_PAGE_SIZE,
    total: 0,
  });

  // 是否还有更多数据
  const hasMore = computed(() => {
    const total = presets.value.length;
    const totalAll = Number(pagination.value.total);
    return totalAll > 0 && total < totalAll;
  });

  // =====================
  // 计算属性
  // =====================

  // 过滤后的预设列表
  const filteredPresets = computed(() => {
    if (!searchQuery.value) return presets.value;
    const query = searchQuery.value.toLowerCase();
    return presets.value.filter(p => p.preset?.name.toLowerCase().includes(query));
  });

  // 按排序顺序排列的提示项（后端已按顺序返回）
  const sortedPromptItems = computed(() => {
    return [...promptItems.value];
  });

  // =====================
  // 预设相关方法
  // =====================

  // 获取预设列表（替换）
  const fetchPresets = async (page?: number, pageSize?: number) => {
    const requestPage = page ?? pagination.value.page;
    const requestPageSize = pageSize ?? pagination.value.pageSize;

    const response = await presetClient.listPresets({
      page: requestPage,
      pageSize: requestPageSize,
    });

    presets.value = response.presets;
    pagination.value = {
      page: response.page,
      pageSize: response.pageSize,
      total: Number(response.total),
    };

    return response.presets;
  };

  // 获取所有预设（不分页，用于需要全部数据的场景）
  const fetchAllPresets = async () => {
    const response = await presetClient.listPresets({
      page: DEFAULT_PAGE_NUM,
      pageSize: FETCH_ALL_PAGE_SIZE,
    });
    presets.value = response.presets;
    pagination.value = {
      page: DEFAULT_PAGE_NUM,
      pageSize: response.presets.length,
      total: Number(response.total),
    };
    return response.presets;
  };

  // 加载更多（无限滚动用）
  const loadMore = async () => {
    if (loadingMore.value || !hasMore.value) return;
    loadingMore.value = true;
    try {
      const nextPage = pagination.value.page + 1;
      const response = await presetClient.listPresets({
        page: nextPage,
        pageSize: pagination.value.pageSize,
      });
      presets.value.push(...response.presets);
      pagination.value = {
        page: response.page,
        pageSize: response.pageSize,
        total: Number(response.total),
      };
    } catch (error) {
      console.error('加载更多预设失败:', error);
    } finally {
      loadingMore.value = false;
    }
  };

  // 重置分页（重新从第一页加载）
  const resetPagination = async () => {
    pagination.value.page = DEFAULT_PAGE_NUM;
    await fetchPresets(DEFAULT_PAGE_NUM, pagination.value.pageSize);
  };

  // 获取单个预设
  const fetchPreset = async (id: number) => {
    const response = await presetClient.getPreset({ id });
    if (response.preset) {
      const index = presets.value.findIndex(p => p.preset?.id === id);
      if (index >= 0) {
        // 更新列表中的预设信息（保持prompt_len）
        presets.value[index] = {
          $typeName: 'muse.PresetWithPromptLen',
          preset: response.preset.preset,
          promptLen: response.preset.promptItems?.length || 0
        };
      }
      // 同步更新提示项
      if (response.preset.promptItems) {
        promptItems.value = response.preset.promptItems;
      }
    }
    return response.preset;
  };

  // 创建预设
  const createPreset = async (data: {
    name: string;
    temperature: number;
    topP: number;
    topK: number;
    maxTokens: number;
    frequencyPenalty: number;
    presencePenalty: number;
    promptItems?: Array<{
      identifier: PromptItemIdentifier;
      name: string;
      content?: string;
      role: Role;
      isEnabled: boolean;
      injectionPosition: InjectionPosition;
      injectionDepth: number;
      forbidOverrides: boolean;
      sortOrder: number;
    }>;
  }) => {
    await presetClient.createPreset(data);
    // 创建成功后重新获取第一页列表
    pagination.value.page = DEFAULT_PAGE_NUM;
    await fetchPresets(DEFAULT_PAGE_NUM, pagination.value.pageSize);
  };

  // 更新预设
  const updatePreset = async (
    id: number,
    data: {
      name: string;
      temperature: number;
      topP: number;
      topK: number;
      maxTokens: number;
      frequencyPenalty: number;
      presencePenalty: number;
      version: bigint;
    }
  ) => {
    await presetClient.updatePreset({ id, ...data });
    // 更新成功后重新获取该预设的完整信息
    const updatedPreset = await fetchPreset(id);
    return updatedPreset;
  };

  // 删除预设
  const deletePreset = async (id: number) => {
    await presetClient.deletePreset({ id });
    presets.value = presets.value.filter(p => p.preset?.id !== id);
    if (selectedPreset.value?.preset?.id === id) {
      selectedPreset.value = null;
      promptItems.value = [];
    }
  };

  // 设置活跃预设
  const setActivePreset = async (presetId: number) => {
    await presetClient.setActivePreset({ presetId });
  };

  // 导入预设
  const importPreset = async (fileContent: Uint8Array, fileName: string) => {
    const response = await presetClient.importPreset({ fileContent, fileName });
    if (response.preset) {
      // 添加到列表
      presets.value.push({
        $typeName: 'muse.PresetWithPromptLen',
        preset: response.preset.preset,
        promptLen: response.preset.promptItems?.length || 0
      });
    }
    return response.preset;
  };

  // 导出预设
  const exportPreset = async (id: number) => {
    const response = await presetClient.exportPreset({ id });
    return {
      fileContent: response.fileContent,
      fileName: response.fileName
    };
  };

  // 选中预设
  const selectPreset = async (preset: PresetWithPromptLen | null) => {
    if (preset && preset.preset) {
      // 加载完整的预设信息
      const fullPreset = await fetchPreset(preset.preset.id);
      if (fullPreset) {
        selectedPreset.value = fullPreset;
      }
    } else {
      selectedPreset.value = null;
      promptItems.value = [];
    }
  };

  // =====================
  // 提示项相关方法
  // =====================

  // 获取预设的提示项列表
  const fetchPromptItems = async (presetId: number) => {
    const response = await presetClient.listPromptItems({ presetId });
    promptItems.value = response.items;
    return response.items;
  };

  // 添加提示项
  const addPromptItem = async (
    presetId: number,
    data: {
      identifier: PromptItemIdentifier;
      name: string;
      content?: string;
      role: Role;
      isEnabled: boolean;
      injectionPosition: InjectionPosition;
      injectionDepth: number;
      forbidOverrides: boolean;
      sortOrder: number;
    }
  ) => {
    const response = await presetClient.addPromptItem({ presetId, ...data });
    if (response.item) {
      promptItems.value.push(response.item);
    }
    return response.item;
  };

  // 更新提示项
  const updatePromptItem = async (
    id: number,
    data: {
      identifier: PromptItemIdentifier;
      name: string;
      content?: string;
      role: Role;
      isEnabled: boolean;
      injectionPosition: InjectionPosition;
      injectionDepth: number;
      forbidOverrides: boolean;
      sortOrder: number;
    }
  ) => {
    const response = await presetClient.updatePromptItem({ id, ...data });
    if (response.item) {
      const index = promptItems.value.findIndex(item => item.id === id);
      if (index >= 0) {
        promptItems.value[index] = response.item;
      }
    }
    return response.item;
  };

  // 删除提示项
  const deletePromptItem = async (id: number) => {
    await presetClient.deletePromptItem({ id });
    promptItems.value = promptItems.value.filter(item => item.id !== id);
  };

  // 更新提示项排序
  const updatePromptItemsOrder = async (presetId: number) => {
    await presetClient.updatePromptItemsOrder({ presetId });
    // 重新获取提示项列表以更新本地状态
    await fetchPromptItems(presetId);
  };

  // =====================
  // 辅助方法
  // =====================

  // 设置搜索关键词
  const setSearchQuery = (query: string) => {
    searchQuery.value = query;
  };

  // 重置状态
  const reset = () => {
    presets.value = [];
    selectedPreset.value = null;
    promptItems.value = [];
    searchQuery.value = '';
    pagination.value = {
      page: DEFAULT_PAGE_NUM,
      pageSize: DEFAULT_PAGE_SIZE,
      total: 0
    };
  };

  return {
    // 状态
    presets,
    selectedPreset,
    promptItems,
    sortedPromptItems,
    loading,
    loadingMore,
    searchQuery,
    filteredPresets,
    hasMore,

    // 预设方法
    fetchPresets,
    fetchAllPresets,
    loadMore,
    resetPagination,
    fetchPreset,
    createPreset,
    updatePreset,
    deletePreset,
    setActivePreset,
    importPreset,
    exportPreset,
    selectPreset,

    // 提示项方法
    fetchPromptItems,
    addPromptItem,
    updatePromptItem,
    deletePromptItem,
    updatePromptItemsOrder,

    // 辅助方法
    setSearchQuery,
    reset
  };
});
